File: src/main/java/com/github/simiacryptus/aicoder/actions/MarkdownNewTableColsAction.java
package com.github.simiacryptus.aicoder.actions;

import com.github.simiacryptus.aicoder.config.AppSettingsState;
import com.github.simiacryptus.aicoder.openai.CompletionRequest;
import com.github.simiacryptus.aicoder.util.psi.PsiUtil;
import com.github.simiacryptus.aicoder.util.StringTools;
import com.github.simiacryptus.aicoder.util.UITools;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.editor.Caret;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.github.simiacryptus.aicoder.util.UITools.replaceString;

public class MarkdownNewTableColsAction extends AnAction {

    @Override
    public void update(@NotNull AnActionEvent e) {
        e.getPresentation().setEnabledAndVisible(isEnabled(e));
        super.update(e);
    }

    private static boolean isEnabled(@NotNull AnActionEvent e) {
        return null != getMarkdownNewTableColsParams(e);
    }

    @Nullable
    public static MarkdownNewTableColsParams getMarkdownNewTableColsParams(@NotNull AnActionEvent e) {
        Caret caret = e.getData(CommonDataKeys.CARET);
        if (null != caret) {
            PsiFile psiFile = e.getData(CommonDataKeys.PSI_FILE);
            if (null != psiFile) {
                PsiElement table = PsiUtil.getSmallestIntersecting(psiFile, caret.getSelectionStart(), caret.getSelectionEnd(), "MarkdownTableImpl");
                if (null != table) {
                    List<CharSequence> rows = Arrays.asList(StringTools.transposeMarkdownTable(PsiUtil.getAll(table, "MarkdownTableRowImpl")
                            .stream().map(PsiElement::getText).collect(Collectors.joining("\n")), false, false).split("\n"));
                    CharSequence n = Integer.toString(rows.size() * 2);
                    return new MarkdownNewTableColsParams(caret, table, rows, n);
                }
            }
        }
        return null;
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent event) {
        MarkdownNewTableColsParams markdownNewTableColsParams = getMarkdownNewTableColsParams(event);
        AppSettingsState settings = AppSettingsState.getInstance();
        CharSequence indent = UITools.getIndent(Objects.requireNonNull(markdownNewTableColsParams).caret);
        CompletionRequest request = MarkdownNewTableColAction.newRowsRequest(settings, markdownNewTableColsParams.n, markdownNewTableColsParams.rows, "");
        UITools.redoableRequest(request, "", event, newText -> transformCompletion(markdownNewTableColsParams, indent, newText), newText -> replaceString(event.getRequiredData(CommonDataKeys.EDITOR).getDocument(),
                markdownNewTableColsParams.table.getTextRange().getStartOffset(),
                markdownNewTableColsParams.table.getTextRange().getEndOffset(), newText));
    }

    @NotNull
    private static String transformCompletion(MarkdownNewTableColsParams markdownNewTableColsParams, CharSequence indent, CharSequence complete) {
        List<CharSequence> newRows = Arrays.stream(("" + complete).split("\n")).map(String::trim)
                .filter(x -> x.length() > 0).collect(Collectors.toList());
        String newTableTxt = StringTools.transposeMarkdownTable(Stream.concat(markdownNewTableColsParams.rows.stream(), newRows.stream())
                .collect(Collectors.joining("\n")), false, true);
        return newTableTxt.replace("\n", "\n" + indent);
    }

    public static class MarkdownNewTableColsParams {
        public final Caret caret;
        public final PsiElement table;
        public final List<CharSequence> rows;
        public final CharSequence n;

        private MarkdownNewTableColsParams(Caret caret, PsiElement table, List<CharSequence> rows, CharSequence n) {
            this.caret = caret;
            this.table = table;
            this.rows = rows;
            this.n = n;
        }

    }
}

File: src/main/kotlin/com/github/simiacryptus/aicoder/actions/markdown/MarkdownNewTableColsAction.kt
package com.github.simiacryptus.aicoder.actions.markdown

import com.github.simiacryptus.aicoder.config.AppSettingsState
import com.github.simiacryptus.aicoder.util.ComputerLanguage
import com.github.simiacryptus.aicoder.util.StringTools
import com.github.simiacryptus.aicoder.util.UITools
import com.github.simiacryptus.aicoder.util.psi.PsiUtil
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.editor.Caret
import com.intellij.psi.PsiElement
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

class MarkdownNewTableColsAction : AnAction() {
    override fun update(e: AnActionEvent) {
        e.presentation.isEnabledAndVisible = isEnabled(e)
        super.update(e)
    }

    override fun actionPerformed(event: AnActionEvent) {
        val markdownNewTableColsParams = getMarkdownNewTableColsParams(event)
        val settings = AppSettingsState.getInstance()
        val indent = UITools.getIndent(Objects.requireNonNull(markdownNewTableColsParams)!!.caret)
        val request = MarkdownNewTableColAction.newRowsRequest(
            settings,
            markdownNewTableColsParams!!.n,
            markdownNewTableColsParams.rows,
            ""
        )
        UITools.redoableRequest(request, "", event,
            { newText ->
                transformCompletion(
                    markdownNewTableColsParams, indent, newText
                )
            },
            { newText ->
                UITools.replaceString(
                    event.getRequiredData(CommonDataKeys.EDITOR)
                        .document,
                    markdownNewTableColsParams.table.textRange.startOffset,
                    markdownNewTableColsParams.table.textRange.endOffset, newText!!
                )
            }
        )
    }

    class MarkdownNewTableColsParams constructor(
        val caret: Caret,
        val table: PsiElement,
        val rows: List<CharSequence>,
        val n: CharSequence
    )

    companion object {
        private fun isEnabled(e: AnActionEvent): Boolean {
            if(UITools.isSanctioned()) return false
            val computerLanguage = ComputerLanguage.getComputerLanguage(e) ?: return false
            return if (ComputerLanguage.Markdown !== computerLanguage) false else null != getMarkdownNewTableColsParams(
                    e
                )
        }

        fun getMarkdownNewTableColsParams(e: AnActionEvent): MarkdownNewTableColsParams? {
            val caret = e.getData(CommonDataKeys.CARET)
            if (null != caret) {
                val psiFile = e.getData(CommonDataKeys.PSI_FILE)
                if (null != psiFile) {
                    val table =
                        PsiUtil.getSmallestIntersecting(
                            psiFile,
                            caret.selectionStart,
                            caret.selectionEnd,
                            "MarkdownTableImpl"
                        )
                    if (null != table) {
                        val rows =
                            Arrays.asList<CharSequence>(*StringTools.transposeMarkdownTable(
                                PsiUtil.getAll(table, "MarkdownTableRowImpl")
                                    .stream().map { obj: PsiElement -> obj.text }.collect(Collectors.joining("\n")),
                                false,
                                false
                            ).split("\n".toRegex()).dropLastWhile { it.isEmpty() }
                                .toTypedArray())
                        val n: CharSequence = Integer.toString(rows.size * 2)
                        return MarkdownNewTableColsParams(caret, table, rows, n)
                    }
                }
            }
            return null
        }

        private fun transformCompletion(
            markdownNewTableColsParams: MarkdownNewTableColsParams,
            indent: CharSequence,
            complete: CharSequence?
        ): String {
            val newRows = Arrays.stream(("" + complete).split("\n".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()).map { obj: String -> obj.trim { it <= ' ' } }
                .filter { x: String -> x.length > 0 }.collect(Collectors.toList())
            val newTableTxt = StringTools.transposeMarkdownTable(
                Stream.concat(markdownNewTableColsParams.rows.stream(), newRows.stream())
                    .collect(Collectors.joining("\n")), false, true
            )
            return newTableTxt.replace("\n", "\n$indent")
        }
    }
}